controllers/workflow_controller.go

```diff
@@ -186,37 +167,45 @@ func (r *WorkflowReconciler) Reconcile(ctx context.Context, req ctrl.Request) (r
 		}
 	}
 
-	driverDone, err := checkDriverStatus(workflow)
-	if err != nil {
-		if workflow.Status.Status != "Error" {
-			log.Info("Workflow state transitioning to Error")
+	// If the workflow has already been marked as complete for this state, then
+	// we don't need to check the drivers. The drivers can't transition from complete
+	// to not complete
+	if workflow.Status.Ready == true {
+		return ctrl.Result{}, nil
+	}
+
+	workflow.Status.Ready = true
+	workflow.Status.Status = "Completed"
+	workflow.Status.Message = ""
+
+	// Loop through the driver status array and update the workflow
+	// status as necessary
+	for _, driver := range workflow.Status.Drivers {
+		if driver.WatchState != workflow.Status.State {
+			continue
 		}
 
-		workflow.Status.Status = "Error"
-		workflow.Status.Message = err.Error()
-	} else {
-		// Set Ready/Status based on driverDone condition
-		// All drivers achieving the current desiredStatus means we've achieved the desired state
-		if driverDone == ConditionTrue {
-			if workflow.Status.Ready != ConditionTrue {
-				ts := metav1.NowMicro()
-				workflow.Status.ReadyChange = &ts
-				workflow.Status.ElapsedTimeLastState = ts.Time.Sub(workflow.Status.DesiredStateChange.Time).Round(time.Microsecond).String()
-				workflow.Status.Status = "Completed"
-				workflow.Status.Message = "Workflow " + workflow.Status.State + " completed successfully"
-				log.Info("Workflow transitioning to ready state " + workflow.Status.State)
-			}
-		} else {
-			// Driver not ready, update Status if not already in DriverWait
-			if workflow.Status.Status != "DriverWait" {
-				workflow.Status.Status = "DriverWait"
-				workflow.Status.Message = "Workflow " + workflow.Status.State + " waiting for driver completion"
-				log.Info("Workflow state=" + workflow.Status.State + " waiting for driver completion")
-			}
+		if driver.Completed == false {
+			workflow.Status.Ready = false
+			workflow.Status.Status = "DriverWait"
+		}
+
+		if driver.Message != "" {
+			workflow.Status.Message = driver.Message
```

Comment:
We should be able to clear workflow.Status.Message when we move to the next state.  But within the state, we should keep any message that is already in there.

Reply:
I don't think it makes sense to keep the message in the workflow.Status.Message field after the driver has removed the message from the driver array entry. The message might be something like "Could not allocate space for file system". If that error doesn't persist, and the driver has cleared its message field, then the workflow.Status.Message field should be cleared too.